Tests/NIOSSHTests/SSHConnectionStateMachineTests.swift

```diff
@@ -512,6 +512,78 @@ final class SSHConnectionStateMachineTests: XCTestCase {
+        let plaintext = TinyEncryptionAlgorithm.decrypt(key: key, ciphertext: ciphertext)
+        XCTAssertEqual(message, plaintext)
+        print(message.readableBytesView.map { $0 })
+        print(plaintext.readableBytesView.map { $0 })
+    }
+
+    func testTestTransportProtection() throws {
+        let inboundEncryptionKey = SymmetricKey(size: .bits128)
+        let outboundEncryptionKey = SymmetricKey(size: .bits128)
+        let inboundMACKey = SymmetricKey(size: .bits128)
+        let outboundMACKey = SymmetricKey(size: .bits128)
+        let client = TestTransportProtection(initialKeys: .init(
+            initialInboundIV: [],
+            initialOutboundIV: [],
+            inboundEncryptionKey: inboundEncryptionKey,
+            outboundEncryptionKey: outboundEncryptionKey,
+            inboundMACKey: inboundMACKey,
+            outboundMACKey: outboundMACKey
+        ))
+        let server = TestTransportProtection(initialKeys: .init(
+            initialInboundIV: [],
+            initialOutboundIV: [],
+            inboundEncryptionKey: outboundEncryptionKey,
+            outboundEncryptionKey: inboundEncryptionKey,
+            inboundMACKey: outboundMACKey,
+            outboundMACKey: inboundMACKey
+        ))
+        let message = SSHMessage.channelRequest(.init(recipientChannel: 1, type: .exec("uname"), wantReply: false))
+        let allocator = ByteBufferAllocator()
+        var buffer = allocator.buffer(capacity: 1024)
+        XCTAssertNoThrow(try client.encryptPacket(.init(message: message), to: &buffer))
+        XCTAssertNoThrow(try server.decryptFirstBlock(&buffer))
+        var decoded = try server.decryptAndVerifyRemainingPacket(&buffer)
+        XCTAssertEqual(message, try decoded.readSSHMessage())
+    }
+
+    func testReceivingChannelMessagesGetForwardedOnceConnectionMadeUsingTinyEncryptionDripFeedingBytes() throws {
+        let allocator = ByteBufferAllocator()
+        let loop = EmbeddedEventLoop()
+        let schemes: [NIOSSHTransportProtection.Type] = [TestTransportProtection.self]
+        var client = SSHConnectionStateMachine(role: .client(.init(userAuthDelegate: InfinitePasswordDelegate(), serverAuthDelegate: AcceptAllHostKeysDelegate())), protectionSchemes: schemes)
+        var server = SSHConnectionStateMachine(role: .server(.init(hostKeys: [NIOSSHPrivateKey(ed25519Key: .init())], userAuthDelegate: DenyThenAcceptDelegate(messagesToDeny: 0))), protectionSchemes: schemes)
+        try assertSuccessfulConnection(client: &client, server: &server, allocator: allocator, loop: loop)
+
+        let message = SSHMessage.channelData(.init(recipientChannel: 1, data: ByteBuffer(repeating: 17, count: 5)))
+
+        var tempBuffer = allocator.buffer(capacity: 1024)
+        XCTAssertNoThrow(try client.processOutboundMessage(message, buffer: &tempBuffer, allocator: allocator, loop: loop))
+        XCTAssert(tempBuffer.readableBytes > 0)
+
+        while var next = tempBuffer.readSlice(length: 1) {
```

Comment:
the only alternative I was able to come up so far would be to expose state machine state and check if first block in the buffer is decoded, not sure if it's better

Reply:
I don't think exposing the state machine state is the right approach here. This is probably fine.